### DISCUSSION TITLE
Can the constrain cache be global?

### DISCUSSION THREAD

**Discussion by michallepicki** [LPTK/simple-sub#46] [General]
I ported the gist of the algorithm to Erlang ([my repo](https://github.com/michallepicki/suber)) and started wondering - the paper and [comments in code](https://github.com/LPTK/simple-sub/blob/3d873493bc146f0e3d037f8b454efb983c2f1bcb/shared/src/main/scala/simplesub/Typer.scala#L90) say that the `constrain` method has a cache in order to memoize results of previous subtyping tests. I am no Scala expert but from my quick testing, the `implicit cache` gets initialized to an empty mutable set every time it's called from outside of itself.

Question: Can this cache be global per typing environment? Or would that break the algorithm for some edge cases? Making this change does not seem to break any of the examples included in this repo's tests.

**Comment by michallepicki** [comment]
Here is a Scala snippet that I used to check when such implicit gets initialized:
```scala
import scala.collection.mutable
object HelloWorld {
   def main(args: Array[String]) {
      run(0, true)
      run(0, true)
   }
   
   def run(arg: Int, callAgain: Boolean)(implicit cache: mutable.Set[Int] = mutable.Set.empty) {
       println(cache)
       cache += arg
       if (callAgain) {
           run(arg + 1, false)
       }
   }
}
```
Result:
```
$scalac *.scala
$scala HelloWorld
Set()
Set(0)
Set()
Set(0)
```

**Comment by LPTK** [comment]
Hi @michallepicki, cool project, thanks for sharing!

The constrain method could memoize everything across top-level `constrain` calls by making the cache global, but that would essentially constitute a pretty bad space leak, as many type variables are generated during type checking and you'd quickly fill the cache with types involving these variables. But making the cache "local" to each typing environment, as you seem to suggest, would probably be fine. I'm not sure it'd make a big difference, though.

Another thing that could be done would be to cache things according to their generalization level. So there would be a global cache for global types (those at level 0), and one additional cache per nested level. Again, not sure it's worth the over-engineering!
